services/motion/builtin/builtin.go

```diff
@@ -237,71 +247,191 @@ func (ms *builtIn) MoveOnGlobe(
+	var cancelFn context.CancelFunc
+	defer func() {
+		if cancelFn != nil {
+			cancelFn()
+		}
+	}()
+
+	// helper function to manage polling functions
+	startPolling := func(ctx context.Context, period time.Duration, fn func(context.Context) error) {
+		backgroundWorkers.Add(1)
+		goutils.ManagedGo(func() {
+			ticker := time.NewTicker(period)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					if err := fn(ctx); err != nil {
+						errChan <- err
+					}
+					return
 				}
+			}
+		}, backgroundWorkers.Done)
+	}
+
+	// start the loop that (re)plans when something is read from the replan channel
+	// and exits when something is read from the success channel
+	for {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		case <-successChan:
+			return true, nil
+		case err := <-errChan:
+			return false, err
+		case <-replanChan:
+			// cancel the goroutines spawned by this function, create a new cancellable context to use in the future
+			if cancelFn != nil {
+				cancelFn()
+			}
+			backgroundWorkers.Wait()
+			// context is not lost because it is cancelled above and then cancelled again with defer
+			//nolint:govet
+			cancelCtx, cancelFn = context.WithCancel(ctx)
+
+			inputs, err := kb.CurrentInputs(ctx)
+			if err != nil {
+				//nolint:govet
+				return false, err
+			}
+			// TODO: this is really hacky and we should figure out a better place to store this information
+			if len(kb.Kinematics().DoF()) == 2 {
+				inputs = inputs[:2]
+			}
+			planRequest.StartConfiguration = map[string][]referenceframe.Input{componentName.Name: inputs}
+
+			plan, err := motionplan.PlanMotion(ctx, planRequest)
+			if err != nil {
 				return false, err
 			}
+
+			// drain channels of any extra messages
+			for len(replanChan) > 0 {
+				<-replanChan
+			}
+			for len(errChan) > 0 {
+				<-errChan
+			}
+
+			// spawn two goroutines that each have the ability to trigger a replan
+			// TODO: optionally could make MotionConfiguration more robust such that the option that these are zero is never possible
+			if positionPollingPeriod > 0 {
+				startPolling(cancelCtx, positionPollingPeriod, func(ctx context.Context) error {
+					// TODO: the function that actually monitors position
+					return nil
+				})
+			}
+			if obstaclePollingPeriod > 0 {
+				startPolling(cancelCtx, obstaclePollingPeriod, func(ctx context.Context) error {
+					// TODO: the function that actually monitors obstacles
+					replanChan <- true
+					return nil
+				})
+			}
+
+			// spawn function to execute the plan on the robot
+			backgroundWorkers.Add(1)
+			goutils.ManagedGo(func() {
+				if err := ms.executePlan(cancelCtx, kb, plan); err != nil {
+					errChan <- err
+					return
+				}
+
+				// the plan has been fully executed so check to see if the GeoPoint we are at is close enough to the goal.
+				success, err := arrivedAtGoal(cancelCtx, movementSensor, destination, motionCfg.PlanDeviationMM)
```

Comment:
If we don't arrive at the goal & don't have an error, won't we deadlock?

Reply:
Good point, I think we will.
We may want to send a `replanChan <- true` in the case of a `false, nil` return here.